testsuite/tests/typing-jkind-bounds/no-infer-across-modules/test.ml

```diff
@@ -0,0 +1,275 @@
+|}]
+
+let foo (t : int rec1 @@ contended) = use_uncontended t
+[%%expect{|
+val foo : int Define_with_kinds.rec1 @ contended -> unit = <fun>
+|}]
+
+let foo (t : int my_ref rec1 @@ contended) = use_uncontended t
+[%%expect{|
+Line 1, characters 61-62:
+1 | let foo (t : int my_ref rec1 @@ contended) = use_uncontended t
+                                                                 ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+let foo (t : int ref rec1 @@ contended) = use_uncontended t
+[%%expect{|
+Line 1, characters 58-59:
+1 | let foo (t : int ref rec1 @@ contended) = use_uncontended t
+                                                              ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+(******)
+
+let use_uncontended_three_values : ('a : (value & value) & value) @ uncontended -> unit = fun _ -> ()
+let use_portable_three_values : ('a : (value & value) & value) @ portable -> unit = fun _ -> ()
+
+let foo (t : #(int * int) unboxed_rec @@ contended) = use_uncontended_three_values t
+[%%expect{|
+val use_uncontended_three_values : ('a : (value & value) & value). 'a -> unit =
+  <fun>
+val use_portable_three_values :
+  ('a : (value & value) & value). 'a @ portable -> unit = <fun>
+val foo : #(int * int) Define_with_kinds.unboxed_rec @ contended -> unit =
+  <fun>
+|}]
+
+let foo (t : #(int my_ref * int) unboxed_rec @@ contended) = use_uncontended_three_values t
+[%%expect{|
+Line 1, characters 90-91:
+1 | let foo (t : #(int my_ref * int) unboxed_rec @@ contended) = use_uncontended_three_values t
+                                                                                              ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+let foo (t : #((int -> int) * int) unboxed_rec @@ nonportable) = use_portable_three_values t
+[%%expect{|
+Line 1, characters 91-92:
+1 | let foo (t : #((int -> int) * int) unboxed_rec @@ nonportable) = use_portable_three_values t
+                                                                                               ^
+Error: This value is "nonportable" but expected to be "portable".
+|}, Principal{|
+Line 1, characters 91-92:
+1 | let foo (t : #((int -> int) * int) unboxed_rec @@ nonportable) = use_portable_three_values t
+                                                                                               ^
+Error: This expression has type
+         "#((int -> int) * int) Define_with_kinds.unboxed_rec"
+       but an expression was expected of type "('a : (value & value) & value)"
+       The kind of #((int -> int) * int) Define_with_kinds.unboxed_rec is
+         (immediate & immediate) & immediate.
+       But the kind of #((int -> int) * int) Define_with_kinds.unboxed_rec must be a subkind of
+         (value & value) & value
+         because of the definition of use_portable_three_values at line 2, characters 84-95.
+|}]
+
+(******)
+
+module Option : Optionish = struct
```

Comment:
I'd also like to see:
```
module rec Foo : Optionish = Foo
```
```
type 'a my_option = 'a Optionish2.my_option = ...
```
Maybe there's some others along this line I'm not thinking of?